add implementation for Path2D addPath method

arthmis opened the pull request:
Add implementation for Path2D addPath method
Spec: https://html.spec.whatwg.org/multipage/canvas.html#dom-path2d-addpath

Testing: WPT test - `tests/wpt/tests/css/geometry/DOMMatrix2DInit-validate-fixup.html`
Fixes: #37695

arthmis committed:
add  initial functionality for add_path overload

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
implement transformations on path segments

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
update idl for addPath

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
update webidl for canvas rendering to have one addPath method and update
implementation of it for Path2D

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add most of the implementation for step 6 and 8 of addPath

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add implementation for last point and complete addPath steps

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
refactor last_point function

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented on components/script_bindings/webidls/CanvasRenderingContext2D.webidl:
```diff
@@ -287,6 +287,7 @@ interface Path2D {
   constructor(DOMString pathString);
-  undefined addPath(Path2D path/*, SVGMatrix? transformation*/);
+  [Throws]
+  undefined addPath(Path2D path, optional DOMMatrix2DInit transform = {});
```
I know this is technically the correct definition for this but should there also be another definition like
```
undefined addPath(Path2D path);
```
with an implementation that skips applying the transformation on the `PathSegments`? Or should I skip applying the matrix if it's the identity matrix?

sagudev replied:
Honestly, I would just keep multiplying identity matrix for simplicity.

arthmis commented:
@sagudev I'm not ready for review yet but just had some questions. I've left them in the code comments.

arthmis commented on components/script/dom/path2d.rs:
```diff
@@ -59,20 +63,198 @@ impl Path2D {
+        self.path.borrow().is_empty()
+    }
+
+    fn last_point(path: &[PathSegment]) -> Option<PathSegment> {
```
for this function, since I'm not working with points but `PathSegment` is it ok to just return `PathSegment::ClosePath` if a last point can't be determined? For example, there are only `ClosePath` segments found in the path.

sagudev replied:
Shouldn't this function return a Option<Point2D>? And for all closepaths it should return None (as per spec if subpath is empty closepath should be nop, so we should not even insert it).

arthmis replied:
Did you mean this part?
```rust
let apply_matrix_transform = |x: f32, y: f32| -> Point2D<f64> {
    matrix.transform_point(Point2D::new(x.into(), y.into()))
};
```
If so, after you made the changes to `dommatrix2dinit_to_matrix` function the `Option` wasn't necessary anymore.

arthmis replied:
Oh I didn't return a Point2d because we don't store a list of points but `PathSegment` so I just returned the appropriate `PathSegment` or None.

arthmis replied:
I'm hitting some difficulties with `PathSegment::ArcTo`. From the spec `ArcTo` should use the previous point as its starting point and if there isn't a previous point then it uses the first control point as the starting. So in the situation where a path looks like `[ArcTo, ArcTo, ArcTo]` and I want to get the last point for such a subpath, do I have compute the point generated by the 2nd `ArcTo` in that array in order to use it as the previous point for the last `ArcTo`?

sagudev replied:
Unfortunately, yes.

arthmis commented on components/script/dom/path2d.rs:
```diff
@@ -59,20 +63,198 @@ impl Path2D {
+        let mut c = other.segments();
+
+        // Step 5. Transform all the coordinates and lines in c by the transform matrix matrix.
+        let apply_matrix_transform = |x: f32, y: f32| -> Point2D<f64> {
```
The matrix uses `f64` for its calculations, should we maybe change the `PathSegment` to contain f64 instead? I had to do lossy conversion from `f64` to `f32` after calculating the transformation.

sagudev replied:
Probably yes, but not in this PR.

arthmis committed:
only push last point if it is a pathsegment that isn't ClosePath

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
remove redundant empty check

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented:
@sagudev I'm trying to run the WPT for the path2d functionality found here `tests\wpt\tests\html\canvas\element\path-objects` in the repo. They all crash though, is that expected at this stage? I use this command `./mach test-wpt tests\wpt\tests\html\canvas\element\path-objects -r`.

I also get a message of `Unrecognized option: 'enable-experimental-web-platform-features'`. I'm on Windows 10 if that matters. I also noticed that most of the tests don't have corresponding `ini` files in this meta folder `tests\wpt\meta\html\canvas\element\path-objects`. Do I need to add the `ini` files?

arthmis committed:
update DOMMatrix2DInit-validate-fixup.html.ini to expect PASS when
running test

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis commented:
> Not sure about status of WPT runs on windows, but WPT runs in CI should work (uses ubuntu runners): https://book.servo.org/hacking/testing.html?highlight=mach%20try#running-web-platform-tests-on-your-github-fork

Seems running the WPT tests on windows doesn't work out of the box. I had to switch to my macbook for the tests to run.

sagudev commented on tests/wpt/meta/css/geometry/DOMMatrix2DInit-validate-fixup.html.ini:
Did you manually make this changes, because usually when doing `mach update-wpt` PASS results are omitted.

arthmis commented on components/script/dom/path2d.rs:
```diff
@@ -59,20 +63,195 @@ impl Path2D {
+
+        // Step 8. Create a new subpath in `a` with (x, y) as the only point in the subpath.
+        if let Some(last_point) = last_point {
+            self.push(last_point);
```
Ok, so for a `ArcTo` pathsegment what should I return as the point? It only has control points unlike the other pathsegments.

arthmis commented on components/script/dom/path2d.rs:
```diff
@@ -59,20 +63,195 @@ impl Path2D {
+                        y: transformed_point.y as f32,
+                        radius_x,
+                        radius_y,
+                        rotation,
```
should rotation also be transformed, the spec only says transform coordinates and lines

sagudev replied:
Ah, sorry didn't see this. Yes, apparently we do not need this.

arthmis replied:
nice

arthmis committed:
run ./mach update-wpt for now passing tests

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add implementation to compute last point when adding a path

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add some test cases when computing last point and updating
implementation to fix test cases that fail

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
update ClosePath implementation to be closer to the spec by preventing
insertion of ClosePath segments in an empty path or having consecutive
ClosePath segments

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

arthmis committed:
add more test cases and adjust implementation to pass tests

- remove unnecessary comments and commented code - remove dead code

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

sagudev commented on components/script/dom/path2d.rs:
```diff
@@ -59,24 +63,342 @@ impl Path2D {
+        match last_point {
+            PathSegment::MoveTo { x, y } |
+            PathSegment::LineTo { x, y } |
+            PathSegment::Quadratic { x, y, .. } |
+            PathSegment::Bezier { x, y, .. } |
+            PathSegment::Ellipse { x, y, .. } |
+            PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
```
shouldn't we use last xy pair here, given that we search for last not first point?

sagudev replied:
My bad, I forgot how destruct worked in pattern matching, xy are always of endpoint: `PathSegment::Quadratic { cpx, cpy, x, y }` and all others are names control points.

arthmis committed:
address nit

Signed-off-by: Lloyd Massiah <artmis9@protonmail.com>

sagudev commented on components/script/dom/path2d.rs:
```diff
@@ -59,24 +63,344 @@ impl Path2D {
+
+                let first_point = path.get(first_point_index)?;
+                match first_point {
+                    PathSegment::MoveTo { x, y } |
+                    PathSegment::LineTo { x, y } |
+                    PathSegment::Quadratic { x, y, .. } |
+                    PathSegment::Bezier { x, y, .. } |
+                    PathSegment::Ellipse { x, y, .. } |
+                    PathSegment::SvgArc { x, y, .. } => Some(Point2D::new(*x, *y)),
```
Although to continue my point from https://github.com/servo/servo/pull/37838/files#r2195651271, shouldn't we use first (control) points here?`

arthmis replied:
No at least according to the spec, all of them state to connect the previous point to given point. For example [quadratic curve to](https://html.spec.whatwg.org/multipage/canvas.html#dom-context-2d-quadraticcurveto) says that as well as `bezierCurve`.

Oh actually, `SvgArc` and `Ellipse` might need their own computation. It seems the `x, y` pair isn't necessarily the end point of the computed arc.

arthmis replied:
After looking back at this, I think we should ignore `SvgArc` because it doesn't look like it's a method used for `Path2D`. I left a question about `Ellipse` on a line above this. With the exception of these segments all the segments should use the `x, y` pair for the new subpath after `closePath` is called.

arthmis replied:
ahhh ok, good to know, thanks

sagudev commented on components/script/dom/path2d.rs:
```diff
@@ -59,24 +63,344 @@ impl Path2D {
+                    PathSegment::LineTo { x, y } |
+                    PathSegment::Quadratic { x, y, .. } |
+                    PathSegment::Bezier { x, y, .. } |
+                    PathSegment::Ellipse { x, y, .. } |
```
Indeed, that means we need to look for endpoint of previous subpath. Ugh.

sagudev commented:
With https://github.com/servo/servo/pull/37967 landing soon, we will not need to deal with transforming PathSegmnts anymore, as we will switch to kurbo::BezPath for this, which already implements transform. That will simplify this PR greatly.

At the time doing transform on PathSegments seemed easier, but with complications that arose in this PR I saw I was wrong. I wish I knew before that using kurbo::BezPath would be easier.

arthmis committed:
make changes to use kurbo::BezPath instead of the removed PathSegment

Signed-off-by: arthmis <artmis9@protonmail.com>

sagudev commented on components/script/dom/path2d.rs:
```diff
@@ -48,14 +53,94 @@ impl Path2D {
+        }
+
+        // Step 6. Let (x, y) be the last point in the last subpath of c
+        let last_point = self.path.borrow_mut().last_point();
```
Instead of doing borrow_mut multiple times, we should do it once and use this ref.

arthmis committed:
make tweak to comment

Signed-off-by: arthmis <artmis9@protonmail.com>

sagudev commented on components/script/dom/path2d.rs:
```diff
@@ -175,3 +260,293 @@ impl Path2DMethods<crate::DomTypeHolder> for Path2D {
 }
+
+#[cfg(test)]
+mod test {
```
I think we can remove this tests as we are relying on kurbo's implementation here. We could also upstream those tests to kurbo.

arthmis commented:
@sagudev sorry this took a bit, I had to pause working on this. I've updated `add_path` to use `kurbo::BezPath`. I think the implementation was straightforward but pls let me know if I've missed anything glaring.

Oh actually, what did you mean by `BezPath` already uses transform?

arthmis committed:
remove unnecessary tests after integrating kurbo::BezPath

Signed-off-by: arthmis <artmis9@protonmail.com>

sagudev commented:
> Oh actually, what did you mean by BezPath already uses transform?

Kurbo has https://docs.rs/kurbo/latest/kurbo/struct.BezPath.html#method.apply_affine which applies transform on path.

arthmis commented on components/script/dom/path2d.rs:
```diff
@@ -48,14 +53,94 @@ impl Path2D {
+        let mut c = other.segments();
+
+        // Step 5. Transform all the coordinates and lines in c by the transform matrix matrix.
+        let apply_matrix_transform =
```
ohh nice

arthmis replied:
ok I've made the changes

arthmis committed:
remove redundant borrows and simplify path transformation

Signed-off-by: arthmis <artmis9@protonmail.com>

arthmis committed:
fix some lint warnings

Signed-off-by: arthmis <artmis9@protonmail.com>
